File: frontend/src/lib/schemas/usage-adoption.ts
import { z } from "zod";

export const usageAdoptionSummarySchema = z.object({
  active_users: z.number().int(),
  new_users: z.number().int(),
  returning_users: z.number().int(),
  questions_asked: z.number().int(),
  documents_uploaded: z.number().int(),
  collections_used: z.number().int(),
  connectors_used: z.number().int(),
  citation_clicks: z.number().int(),
  trust_panel_opens: z.number().int(),
  feedback_submitted: z.number().int(),
  saved_answers: z.number().int(),
  onboarding_completion_rate: z.number().nullable(),
  invitations_sent: z.number().int(),
  invitations_accepted: z.number().int(),
  generated_at: z.string(),
});

export const activationFunnelStepSchema = z.object({
  step: z.string(),
  label: z.string(),
  users_reached: z.number().int(),
  drop_off_rate: z.number().nullable(),
});

export const activeUsersPointSchema = z.object({
  date: z.string(),
  active_users: z.number().int(),
});

export const questionsPerUserBucketSchema = z.object({
  bucket: z.string(),
  user_count: z.number().int(),
});

export const roleAdoptionRowSchema = z.object({
  role: z.string(),
  user_count: z.number().int(),
  active_users: z.number().int(),
  questions_asked: z.number().int(),
  activation_rate: z.number().nullable(),
});

export const usageAdoptionChartsSchema = z.object({
  active_users_series: z.array(activeUsersPointSchema),
  questions_per_user: z.array(questionsPerUserBucketSchema),
  feature_usage: z.record(z.string(), z.number().int()),
  funnel: z.array(activationFunnelStepSchema),
  role_adoption_comparison: z.array(roleAdoptionRowSchema),
  drop_off_points: z.array(activationFunnelStepSchema),
  generated_at: z.string(),
});

export const onboardingStatusSchema = z.enum([
  "not_started",
  "in_progress",
  "completed",
]);

export const usageAdoptionUserRowSchema = z.object({
  user_id: z.string(),
  name: z.string(),
  email: z.string(),
  role: z.string(),
  last_active_at: z.string().nullable().optional(),
  questions_asked: z.number().int(),
  sources_used: z.number().int(),
  citation_clicks: z.number().int(),
  feedback_submitted: z.number().int(),
  saved_answers: z.number().int(),
  onboarding_status: onboardingStatusSchema,
});

export const usageAdoptionUserListSchema = z.object({
  rows: z.array(usageAdoptionUserRowSchema),
  total: z.number().int(),
  page: z.number().int(),
  page_size: z.number().int(),
});

export const onboardingReminderResultSchema = z.object({
  sent: z.boolean(),
});

export type UsageAdoptionSummary = z.infer<typeof usageAdoptionSummarySchema>;
export type ActivationFunnelStep = z.infer<typeof activationFunnelStepSchema>;
export type UsageAdoptionCharts = z.infer<typeof usageAdoptionChartsSchema>;
export type OnboardingStatus = z.infer<typeof onboardingStatusSchema>;
export type UsageAdoptionUserRow = z.infer<typeof usageAdoptionUserRowSchema>;
export type UsageAdoptionUserList = z.infer<typeof usageAdoptionUserListSchema>;
export type OnboardingReminderResult = z.infer<
  typeof onboardingReminderResultSchema
>;

export type UsageAdoptionQuery = {
  from?: string;
  to?: string;
  role?: string;
  page?: number;
  page_size?: number;
};

File: backend/app/domains/admin/schemas/usage_adoption.py
"""Pydantic schemas for the admin usage & adoption report API (F353).

Returned by GET /admin/usage-adoption/*. Aggregates existing usage-event,
chat, citation, feedback, verified-answer, share, and invitation signals
already tracked across prior features into adoption metrics, an activation
funnel, charts, and a per-user table. No new persisted fields are
introduced by this feature.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

OnboardingStatus = str  # "not_started" | "in_progress" | "completed"


class UsageAdoptionSummaryResponse(BaseModel):
    """Org-scoped counters for the top-of-dashboard metric cards."""

    active_users: int = 0
    new_users: int = 0
    returning_users: int = 0
    questions_asked: int = 0
    documents_uploaded: int = 0
    collections_used: int = 0
    connectors_used: int = 0
    citation_clicks: int = 0
    trust_panel_opens: int = 0
    feedback_submitted: int = 0
    saved_answers: int = 0
    onboarding_completion_rate: float | None = None
    invitations_sent: int = 0
    invitations_accepted: int = 0
    generated_at: datetime


class ActivationFunnelStepResponse(BaseModel):
    step: str
    label: str
    users_reached: int
    drop_off_rate: float | None = None


class ActiveUsersPoint(BaseModel):
    date: str
    active_users: int


class QuestionsPerUserBucket(BaseModel):
    bucket: str
    user_count: int


class RoleAdoptionRow(BaseModel):
    role: str
    user_count: int
    active_users: int
    questions_asked: int
    activation_rate: float | None = None


class UsageAdoptionChartsResponse(BaseModel):
    active_users_series: list[ActiveUsersPoint] = Field(default_factory=list)
    questions_per_user: list[QuestionsPerUserBucket] = Field(default_factory=list)
    feature_usage: dict[str, int] = Field(default_factory=dict)
    funnel: list[ActivationFunnelStepResponse] = Field(default_factory=list)
    role_adoption_comparison: list[RoleAdoptionRow] = Field(default_factory=list)
    drop_off_points: list[ActivationFunnelStepResponse] = Field(default_factory=list)
    generated_at: datetime


class UsageAdoptionUserRow(BaseModel):
    user_id: str
    name: str
    email: str
    role: str
    last_active_at: datetime | None = None
    questions_asked: int = 0
    sources_used: int = 0
    citation_clicks: int = 0
    feedback_submitted: int = 0
    saved_answers: int = 0
    onboarding_status: OnboardingStatus = "not_started"


class UsageAdoptionUserListResponse(BaseModel):
    rows: list[UsageAdoptionUserRow] = Field(default_factory=list)
    total: int = 0
    page: int = 1
    page_size: int = 25


class OnboardingReminderResponse(BaseModel):
    sent: bool
